sqlglot/optimizer/canonicalize.py

```diff
@@ -128,17 +132,46 @@ def remove_ascending_order(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
-def _coerce_date(a: exp.Expression, b: exp.Expression) -> None:
+def _coerce_date(
+    a: exp.Expression,
+    b: exp.Expression,
+    promote_to_inferred_datetime_type: bool,
+) -> None:
     for a, b in itertools.permutations([a, b]):
         if isinstance(b, exp.Interval):
             a = _coerce_timeunit_arg(a, b.unit)
+
+        a_type = a.type
         if (
-            a.type
-            and a.type.this in exp.DataType.TEMPORAL_TYPES
+            a_type
+            and a_type.this in exp.DataType.TEMPORAL_TYPES
             and b.type
             and b.type.this in exp.DataType.TEXT_TYPES
         ):
-            _replace_cast(b, exp.DataType.Type.DATETIME)
+            if promote_to_inferred_datetime_type:
+                if b.is_string:
+                    date_text = b.name
+                    if is_iso_date(date_text):
+                        b_type = exp.DataType.Type.DATE
+                    elif is_iso_datetime(date_text):
+                        b_type = exp.DataType.Type.DATETIME
```

Comment:
@barakalon it feels like this `is_iso*` logic is somewhat repeated now. I didn't refactor yet because the other sections where we do it take a `unit` argument as well into account which looks somewhat different, but curious to see if there's room for DRYing it up.